src/runtime/rt0_noos_mips64.go

```diff
@@ -0,0 +1,34 @@
+package runtime
+
+import "unsafe"
+
+const pallocMin = 20 * 1024
+const signalStackSize = 4096
+
+//go:nosplit
+func meminit(freeStart, freeEnd, nodmaStart, nodmaEnd uintptr) {
+	freeSize := freeEnd - freeStart
+	nodmaSize := nodmaEnd - nodmaStart
+	size := freeSize + nodmaSize
+
+	// Estimate the space needed for non-heap allocations
+	palloc := int((freeSize>>(pageShift))*unsafe.Sizeof(emptymspan) + pallocMin)
+
+	// We prefer the non-DMA memory for non-heap objects to preserve as much
+	// as possible of the DMA capable memory for heap allocations
+	pallocInFree := max(palloc-int(nodmaSize), 0)
+
+	// Reduce the arena by the remainder of the non-heap space that did not fit in
+	// the non-DMA memory, properly align the arena
+	arenaSize := freeSize - uintptr(pallocInFree)
+	arenaSize &= ^(uintptr(heapArenaBytes) - 1)
+	arenaStart := freeEnd - arenaSize
+
+	noosMem.free.start = freeStart
+	noosMem.free.end = freeEnd
+	noosMem.nodma.start = nodmaStart
+	noosMem.nodma.end = nodmaEnd
+	noosMem.arenaStart = arenaStart
+	noosMem.arenaSize = arenaSize
+	noosMem.size = size
+}
```

Comment:
I see the point and I think it should be also done for thumb and riscv because the current assembly is unreadable and probably provides no serious saves.

But rt0_* files are by convention all assembly. Rename it to mem_noos_mips64.go. Maybe we can even unify it for all supported archs in mem_noos.go with arch specific constants (if required) in const_*.

Reply:
Yes, I did rewrite this in Go mainly to understand the code. I did put this in mem_noos.go, but I will leave it up to you to migrate the other GOARCHs to utilizing it. I don't want to possibly introduce bugs in other ARCHs with this merge and currently don't have the means to test it properly.